Adding toolbox containers for development and testing

cnsnyder opened the pull request:
Please see the newly added ./containers/README.md for a description of these new fun bits (reproduced below for your convenience):

Subscription Manager & Containers
=================================

This directory contains scripts and configuration files that are necessary for running
subscription-manager in a containerized environment. This environment could be created
using podman. Why do we care about running subscription-manager in a containerized
environment? We believe that it could be useful for testing of subscription-manager.
We also support running and testing of subscription-manager using vagrant, but running
subscription-manager in VM could be a bit cumbersome. Running subscription-manager
using podman could be more flexible for developers too.

Requirements
------------

To create and use a containerized environment it is necessary to have following packages
installed in the system:

 * podman
 * toolbox

   $ sudo dnf install -y podman toolbox

Create a containerized environment
----------------------------------

To create a containerized environment you can simply run following script:

    $ toolbox create -i quay.io/candlepin/subscription-manager:main -c rhsm-dev
    $ toolbox enter rhsm-dev

Reviewing PRs
-------------

Jenkins will build, tag, and push the PR test environment to
quay.io/candlepin/subscription-manager:PR-XXXX where XXXX is the PR number.
This is done as a regular part of the jenkins tests as each test is run in
one of these containers.

This means that for any PR which has had the jenkins tests complete
(regardless of success or failure), you can run the following to reproduce
the jenkins test results (for example if the unit tests failed):

    $ git checkout origin pr/XXXX
    $ toolbox create -i quay.io/candlepin/subscription-manager:PR-XXXX -c
    pr-XXXX
    $ toolbox run -c pr-XXXX ./jenkins/unit.sh

Running the following will let you work like normal from within the environment
    $ toolbox enter pr-XXXX


PS the tests will likely fail on this PR until we get the jenkins nodes updated to have toolbox.

jirihnidek commented on containers/Containerfile:
```diff
@@ -0,0 +1,91 @@
+    python3-coverage \
+    python3-dateutil \
+    python3-decorator \
+    python3-decorator \
```
Please remove duplicity of two same lines

cnsnyder replied:
Will be done on next push!

jirihnidek commented on containers/Containerfile:
```diff
@@ -0,0 +1,91 @@
+WORKDIR /build/subscription-manager
+RUN python3 ./setup.py build
+RUN python3 ./setup.py build_ext --inplace
+RUN yum-builddep -y ./subscription-manager.spec
```
I would run this command before `python3 ./setup.py build`, because it can install some required package that is necessary for building.

cnsnyder replied:
Good point!

jirihnidek commented on containers/Containerfile:
```diff
@@ -0,0 +1,91 @@
+    git \
+    git \
+    glibc-common \
+    gtk3-devel \
```
I hope that we don't need this gkt3 thing anymore!

cnsnyder replied:
I don't think we do, will update!

jirihnidek commented on containers/README.md:
```diff
@@ -0,0 +1,49 @@
+the jenkins test results (for example if the unit tests failed):
+
+    $ git checkout origin pr/XXXX
+    $ toolbox create -i quay.io/candlepin/subscription-manager:PR-XXXX -c
```
I believe that line-break should not be here or please add `\` here.

cnsnyder replied:
Thank you!

jirihnidek commented on containers/README.md:
```diff
@@ -0,0 +1,49 @@
+
+    $ git checkout origin pr/XXXX
+    $ toolbox create -i quay.io/candlepin/subscription-manager:PR-XXXX -c
+    pr-XXXX
```
Shouldn't be here `PR-XXXX` instead `pr-XXXX`?

cnsnyder replied:
Indeed!

jirihnidek commented on containers/README.md:
```diff
@@ -0,0 +1,49 @@
+This means that for any PR which has had the jenkins tests complete
+(regardless of success or failure), you can run the following to reproduce
+the jenkins test results (for example if the unit tests failed):
+
```
I would like to clarify something here. What remote repository is used here? Is it repository of `subscription-manager` or any other?

cnsnyder replied:
So the quay.io project (at quay.io/candlepin/subscription-manager) will be configured to watch for changes in this repository's main branch and to build images from there. The jenkins job also builds and pushes images to the quay.io container repository for each PR.

Once this PR is merged I can update quay to build off main automatically so there is always a base to work from (could also do for any branch we like, mostly useful for those we maintain).

jirihnidek replied:
Thanks :+1:

jirihnidek reviewed (changes requested):
Overall I like this PR. :+1:  I have few comment, questions and comments. I will have probably more later today... :-)

jirihnidek commented on containers/build_and_push.sh:
```diff
@@ -0,0 +1,11 @@
+PROJECT_ROOT=$(git rev-parse --show-toplevel)
+
+pushd $PROJECT_ROOT
+#podman pull quay.io/candlepin/subscription-manager:main
```
When the line is commented, then there should be some explanation, why it is still present or the line should be deleted.

cnsnyder replied:
Thank you!

jirihnidek commented on containers/dev-requirements.txt:
```diff
@@ -0,0 +1,14 @@
+pyinotify
+simplejson
+mock
+Sphinx==1.8.5;python_version<="2.7"
```
Do we need to support Python 2.7?

jirihnidek commented on containers/Containerfile:
```diff
@@ -0,0 +1,91 @@
+RUN pip3 install -r dev-requirements.txt
+RUN git clone https://github.com/candlepin/subscription-manager.git
+WORKDIR /build/subscription-manager
+RUN python3 ./setup.py build
```
Could it be possible to set `python` interpreter in any environment variable? It is currently `python3`, but maybe it will be changed in the future back to `python`.

cnsnyder replied:
Any reason to do that vs updating the Containerfile definition if/when such a change happens in the relevant stable branches?

m-horky replied:
Plain `python` can always lead to ambiguity, in theory. `python3` will always be available as a binary, see [this blogpost](https://developers.redhat.com/blog/2019/05/07/what-no-python-in-red-hat-enterprise-linux-8) for example.

jirihnidek replied:
OK, you both are right. We can leave it as it is ATM.

jirihnidek commented on containers/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+# The version of iniparse in PyPi does not run under Python3.  Fedora has
+# patched it so we have to rely on the RPM.
+# iniparse
+python-dateutil==1.5;python_version<="2.7"
```
Same note about Python 2.7 here.

jirihnidek commented on jenkins/toolbox-run.sh:
```diff
@@ -0,0 +1,12 @@
+PROJECT_ROOT=$(git rev-parse --show-toplevel)
+
+# Run one of our tests inside a newly created toolbox container
+
```
I would like to see here some test of number of command line argument, because we use them later in this bash script. When command line arguments are not used, then the script will produce confusing error messages.

jirihnidek reviewed (changes requested):
I have few more questions and requests.

cnsnyder commented on containers/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+# note there are also rpm packaged deps that aren't
+# setup in pypi or with setup.py
+decorator
+six
+# The version of iniparse in PyPi does not run under Python3.  Fedora has
+# patched it so we have to rely on the RPM.
+# iniparse
+python-dateutil==1.5;python_version<="2.7"
+python-dateutil>=2.0;python_version>="3.0"
```
Yes although I could be convinced that we should not!

m-horky replied:
https://github.com/candlepin/subscription-manager/blob/35efe783bd63c6b660f237969230f73c785733fc/containers/Containerfile#L80
The Containerfile copies them into the container. Would the following work?
```bash
COPY ../requirements.txt ../dev-requirements.txt /build/ 
```

cnsnyder commented on containers/dev-requirements.txt:
```diff
@@ -0,0 +1,14 @@
+pyinotify
+simplejson
+mock
+Sphinx==1.8.5;python_version<="2.7"
+Sphinx>=1.8.5;python_version>="3.0"
```
So it seems. I wonder if it would be worth resurrecting the sphinx docs. On the surface the possibility of them seems useful.
However they seem to be rather broken at the moment with the removal of the gtk bits (as the gtk bits are still referenced in the sphinx docs).

m-horky replied:
I'd rather remove the Sphinx completely. It's reintroduction in the future won't be difficult, the setup can create everything itself. We can use the old config in git history as a reference, but the old generated docs we currently have don't seem to be that useful, making it from a scratch sounds quicker.
We should probably create a card to remove all the Sphinx bits, they have been laying in there untouched for years.

jirihnidek replied:
I agree that we can completely remove this, because nobody uses subscription-manager as Python package and generating  doc string documentation is not useful. We provide D-Bus API, but only for internal use.

cnsnyder commented on Jenkinsfile:
```diff
@@ -4,23 +4,31 @@ pipeline {
+        QUAY_CREDS=credentials('candlepin-quay-bot')
+      }
+      steps {
+        sh readFile(file: 'containers/build_and_push.sh')
```
I believe so! This will be in the next push.

cnsnyder commented:
I think I've addressed all feedback so far! I'm happy to squash the commits after review (or we can just merge with the "squash and merge" button). Thanks again for trying this out and providing feedback!

m-horky commented on containers/Containerfile:
```diff
@@ -0,0 +1,88 @@
+    sudo \
+    vim \
+    wget \
+    xorg-x11-server-Xvfb \
```
We're not testing GUI anymore, I believe this X.Org package can be removed, too?

jirihnidek commented on containers/README.md:
```diff
@@ -0,0 +1,48 @@
+To create a containerized environment you can simply run following script:
+
+    $ toolbox create -i quay.io/candlepin/subscription-manager:main -c rhsm-dev
+    $ toolbox enter rhsm-dev
```
I tried this and get following error. I also used `--verbose` to see some details:

```
[jhnidek@thinkpad-t580 subscription-manager]$ toolbox --verbose create -i quay.io/candlepin/subscription-manager:main -c rhsm-dev
DEBU Running as real user ID 1000                 
DEBU Resolved absolute path to the executable as /usr/bin/toolbox 
DEBU Running on a cgroups v2 host                 
DEBU Checking if /etc/subgid and /etc/subuid have entries for user jhnidek 
DEBU Validating sub-ID file /etc/subuid           
DEBU Validating sub-ID file /etc/subgid           
DEBU TOOLBOX_PATH is /usr/bin/toolbox             
DEBU Migrating to newer Podman                    
DEBU Toolbox config directory is /home/jhnidek/.config/toolbox 
DEBU Current Podman version is 3.2.3              
DEBU Creating runtime directory /run/user/1000/toolbox 
DEBU Old Podman version is 3.2.3                  
DEBU Migration not needed: Podman version 3.2.3 is unchanged 
DEBU Resolving container and image names          
DEBU Container: 'rhsm-dev'                        
DEBU Distribution: ''                             
DEBU Image: 'quay.io/candlepin/subscription-manager:main' 
DEBU Release: ''                                  
DEBU Resolved container and image names           
DEBU Container: 'rhsm-dev'                        
DEBU Image: 'quay.io/candlepin/subscription-manager:main' 
DEBU Release: 'main'                              
DEBU Checking if container rhsm-dev already exists 
DEBU Looking for image quay.io/candlepin/subscription-manager:main 
DEBU Looking for image quay.io/candlepin/subscription-manager:main 
Image required to create toolbox container.
Download quay.io/candlepin/subscription-manager:main (500MB)? [y/N]: y
DEBU Pulling image quay.io/candlepin/subscription-manager:main 
Trying to pull quay.io/candlepin/subscription-manager:main...
Error: Error initializing source docker://quay.io/candlepin/subscription-manager:main: Error reading manifest main in quay.io/candlepin/subscription-manager: manifest unknown: manifest unknown
Error: failed to pull image quay.io/candlepin/subscription-manager:main
```

What can be wrong?

I worked some time ago as you can see in the list of images and containers:

```
[jhnidek@thinkpad-t580 subscription-manager]$ toolbox list
IMAGE ID      IMAGE NAME                                                               CREATED
393dc0352639  quay.io/candlepin/subscription-manager:csnyder_containerized_everything  2 months ago
daf770cfb9ee  registry.fedoraproject.org/fedora-toolbox:34                             2 weeks ago

CONTAINER ID  CONTAINER NAME     CREATED         STATUS      IMAGE NAME
5182edf38c05  fedora-toolbox-34  36 minutes ago  configured  registry.fedoraproject.org/fedora-toolbox:34
dad35270e016  rhsm               2 months ago    exited      quay.io/candlepin/subscription-manager:csnyder_containerized_everything
```

jirihnidek replied:
I was able to create fedora34 container using `toolbox create` this morning. Thus it seems that there is something wrong with our image on quay.io.

cnsnyder replied:
Yea there is no tag pushed for "main" yet. That is probably why

cnsnyder replied:
Try again with the PR review instructions please! That *should* work at this point but it's possible something is messed up.

The tag jenkins created by looking at this pr is "PR-2758" (which didn't exist until jenkins ran to jenkinsfile in this PR)

jirihnidek replied:
Yeah, it makes sense. I can confirm that it works for current PR, when I use:

```
toolbox --verbose create -i quay.io/candlepin/subscription-manager:PR-2758 -c rhsm-dev
```

Thanks

jirihnidek commented on containers/dev-requirements.txt:
```diff
@@ -0,0 +1,13 @@
+mock
+Sphinx>=1.8.5
+git+https://github.com/alikins/pyqver.git#egg=pyqver
+git+https://github.com/awood/nose-xvfb.git#egg=nose-xvfb
```
This is probably not necessary too, because we don't test GUI anymore.

jirihnidek commented on jenkins/toolbox-run.sh:
```diff
@@ -0,0 +1,17 @@
+# Run one of our tests inside a newly created toolbox container
+
+if (( $# != 2)); then
+    >&2 echo -e "This script requires two arguments.\nUsage: $0 [END_OF_TAG] [PATH_TO_SCRIPT_TO_RUN_IN_CONTAINER]\nThis script will create a toolbox container, run the script provided in it, and forcefully destroy the toolbox container."
```
Much more better, but I have one nitpick. Could you split error/warning message into several lines?

cnsnyder replied:
Addressed in latest push :)

jirihnidek reviewed (changes requested):
I have few more comments. I wasn't able to new image/container using toolbox create. More details is in comment. It is possible that I do something wrong.

jirihnidek reviewed (approved):
I approved it, but later I found some small issues. It can be reason for timeout of build from time to time. I will try to help fix it.

jirihnidek commented on containers/build_and_push.sh:
```diff
@@ -0,0 +1,10 @@
+
+pushd $PROJECT_ROOT
+podman build -f ./containers/Containerfile -t quay.io/candlepin/subscription-manager:$CURRENT_GIT_HASH
+podman push --creds $QUAY_CREDS quay.io/candlepin/subscription-manager:$CURRENT_GIT_HASH
```
I can see timout of Jenkins job, then we this script tries to push container to `quay.io`

BTW: why do we try to push container of each commit to the quay? Is it really necessary?

ptoscano replied:
should `$QUAY_CREDS` better be quoted too?

jirihnidek commented on containers/Containerfile:
```diff
@@ -0,0 +1,87 @@
+    desktop-file-utils \
+    dnf-utils \
+    gcc \
+    GConf2-devel \
```
I think that this should not be here too: `GConf2-devel` add lot of dependencies. Packages like `gtk3`, `gtk2` are installed for this reason.

jirihnidek commented on containers/Containerfile:
```diff
@@ -0,0 +1,87 @@
+    cmake \
+    dbus-daemon \
+    dbus-glib \
+    dbus-glib-devel \
```
I believe that this packages was required, because rhsm-icon communicate using D-Bus, but we don't need it anymore, because other compiled tools do not use D-Bus API.

ptoscano commented on containers/build_and_push.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -x
+CURRENT_GIT_HASH=$(git rev-parse HEAD)
+PROJECT_ROOT=$(git rev-parse --show-toplevel)
```
is this a convoluted way for `$PWD` or `pwd`?

ptoscano commented on containers/build_and_push.sh:
```diff
@@ -0,0 +1,10 @@
+CURRENT_GIT_HASH=$(git rev-parse HEAD)
+PROJECT_ROOT=$(git rev-parse --show-toplevel)
+
+pushd $PROJECT_ROOT
```
please quote `$PROJECT_ROOT`, otherwise it will fail if the current path contains a space

ptoscano commented on jenkins/toolbox-run.sh:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash -x
+PROJECT_ROOT=$(git rev-parse --show-toplevel)
```
ditto, convoluted way instead of `$PWD` or `pwd`

cnsnyder replied:
This was done this way so the script could be run from anywhere in the repo. If we were to use $PWD it would only work as expected if run from the right dir.

ptoscano commented on jenkins/toolbox-run.sh:
```diff
@@ -0,0 +1,22 @@
+    exit
+fi
+
+pushd $PROJECT_ROOT
```
ditto, please quote paths

ptoscano commented on containers/dev-requirements.txt:
```diff
@@ -0,0 +1,12 @@
+simplejson
+mock
+Sphinx>=1.8.5
+git+https://github.com/alikins/pyqver.git#egg=pyqver
```
I got rid of pyqver a couple of months ago already...

ptoscano reviewed (commented):
Whew! A lot to go through; sadly I haven't had the time to actually test it yet.

So first, thanks for this, as it seems a potential improvement.

Follows is a list of various notes/questions/issues, in additions to the inline notes left:

- the new shell scripts lack a lot in error handling and recovery: if any of the commands fail, the scripts will continue as if nothing happened... please make sure they can cleanup after themselves in case of issues

- there is a lot of mixup of commits that do something and cleanups afterwards... please squash what is "something and its fixups", so that history browsing, patch checking, and bisecting are usable and working as intended

- some of the changes in PR, e.g. the reorganization of the jenkins scripts and files, are mostly out of scope from "use toolbox & containers for PRs"; they are potential nice changes on their own, and I'd rather review them/accept separately than mixing them in a single big PR

- `containers/dev-requirements.txt` and `containers/requirements.txt` duplicate the content of the equivalent files in the top-level of the repository; can't we please first reorganize/fix the top-level ones in a separate PR (so we can check they don't break the current setups), and then make this container setup use them?

jirihnidek commented on containers/dev-requirements.txt:
```diff
@@ -0,0 +1,10 @@
+-r requirements.txt
+
+nose-capturestderr
+nose-randomly
```
nose-* packages are not needed anymore, because we started to use pytest for unit testing.

jirihnidek reviewed (changes requested):
I have some requests/suggestions related to replacing of `nosetests` with `pytest`.

Lorquas commented:
We need to make a work around until this gets merged: https://github.com/containers/toolbox/pull/867
